File: test/_harnesses/zig/src/multiple.zig
const std = @import("std");
const Checker = @import("./util.zig").Checker;

const small = @import("lib/SmallMessages.zig");

const empty_msg = small.EmptyMessage{};
const byte_msg = small.ByteMessage{ .byteMember = 242 };
const int_msg_a = small.IntMessage{ .intMember = -42 };
const int_msg_b = small.IntMessage{ .intMember = 2048 };
const float_msg = small.FloatMessage{ .floatMember = 1234.5678 };
const long_msg = small.LongMessage{ .intMember = 2147483647 + 10 };

pub fn main() !void {
    const args = try std.process.argsAlloc(std.heap.page_allocator);
    defer std.process.argsFree(std.heap.page_allocator, args);

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const testAllocator = gpa.allocator();

    var checker = Checker{};

    if (std.mem.eql(u8, args[1], "--generate")) {
        var buffer_size: usize = 0;
        buffer_size += byte_msg.getSizeInBytes();
        buffer_size += int_msg_a.getSizeInBytes() * 3;
        buffer_size += int_msg_b.getSizeInBytes() * 4;
        buffer_size += empty_msg.getSizeInBytes() * 2;
        buffer_size += long_msg.getSizeInBytes();
        buffer_size += float_msg.getSizeInBytes();
        buffer_size += 12;

        const buffer = try testAllocator.alloc(u8, buffer_size);
        defer testAllocator.free(buffer);

        var local_offset: usize = 0;
        local_offset += byte_msg.writeBytes(local_offset, buffer, true);
        local_offset += int_msg_a.writeBytes(local_offset, buffer, true);
        local_offset += int_msg_b.writeBytes(local_offset, buffer, true);
        local_offset += empty_msg.writeBytes(local_offset, buffer, true);
        local_offset += long_msg.writeBytes(local_offset, buffer, true);
        local_offset += float_msg.writeBytes(local_offset, buffer, true);
        local_offset += int_msg_a.writeBytes(local_offset, buffer, true);
        local_offset += int_msg_b.writeBytes(local_offset, buffer, true);
        local_offset += int_msg_b.writeBytes(local_offset, buffer, true);
        local_offset += int_msg_b.writeBytes(local_offset, buffer, true);
        local_offset += int_msg_a.writeBytes(local_offset, buffer, true);
        local_offset += empty_msg.writeBytes(local_offset, buffer, true);

        var file = try std.fs.cwd().createFile(args[2], .{ .truncate = true });
        defer file.close();

        try file.writeAll(buffer);
    } else if (std.mem.eql(u8, args[1], "--read")) {
        var file = try std.fs.cwd().openFile(args[2], .{});
        defer file.close();

        const original_buffer = try file.readToEndAlloc(testAllocator, std.math.maxInt(u32));
        defer testAllocator.free(original_buffer);

        const buffer = try testAllocator.alloc(u8, original_buffer.len + 25);
        defer testAllocator.free(buffer);

        std.mem.copyForwards(u8, buffer[0..original_buffer.len], original_buffer);
        for (buffer[original_buffer.len..]) |*b| {
            b.* = 0;
        }

        const msg_list = try small.processRawBytes(testAllocator, buffer);
        defer testAllocator.free(msg_list);
        checker.softAssert(msg_list.len == 12, "reading multiple messages");

        switch (msg_list[0]) {
            .ByteMessage => |bm| checker.softAssert(bm.byteMember == byte_msg.byteMember, "msg 0 content"),
            else => checker.softAssert(false, "msg 0 type"),
        }
        switch (msg_list[1]) {
            .IntMessage => |im| checker.softAssert(im.intMember == int_msg_a.intMember, "msg 1 content"),
            else => checker.softAssert(false, "msg 1 type"),
        }
        switch (msg_list[2]) {
            .IntMessage => |im| checker.softAssert(im.intMember == int_msg_b.intMember, "msg 2 content"),
            else => checker.softAssert(false, "msg 2 type"),
        }
        switch (msg_list[3]) {
            .EmptyMessage => |_| {},
            else => checker.softAssert(false, "msg 3 type"),
        }
        switch (msg_list[4]) {
            .LongMessage => |lm| checker.softAssert(lm.intMember == long_msg.intMember, "msg 4 content"),
            else => checker.softAssert(false, "msg 4 type"),
        }
        switch (msg_list[5]) {
            .FloatMessage => |fm| checker.softAssert(fm.floatMember == float_msg.floatMember, "msg 5 content"),
            else => checker.softAssert(false, "msg 5 type"),
        }
        switch (msg_list[6]) {
            .IntMessage => |im| checker.softAssert(im.intMember == int_msg_a.intMember, "msg 6 content"),
            else => checker.softAssert(false, "msg 6 type"),
        }
        switch (msg_list[7]) {
            .IntMessage => |im| checker.softAssert(im.intMember == int_msg_b.intMember, "msg 7 content"),
            else => checker.softAssert(false, "msg 7 type"),
        }
        switch (msg_list[8]) {
            .IntMessage => |im| checker.softAssert(im.intMember == int_msg_b.intMember, "msg 8 content"),
            else => checker.softAssert(false, "msg 8 type"),
        }
        switch (msg_list[9]) {
            .IntMessage => |im| checker.softAssert(im.intMember == int_msg_b.intMember, "msg 9 content"),
            else => checker.softAssert(false, "msg 9 type"),
        }
        switch (msg_list[10]) {
            .IntMessage => |im| checker.softAssert(im.intMember == int_msg_a.intMember, "msg 10 content"),
            else => checker.softAssert(false, "msg 10 type"),
        }
        switch (msg_list[11]) {
            .EmptyMessage => |_| {},
            else => checker.softAssert(false, "msg 11 type"),
        }
    }

    const mem_result = gpa.deinit();
    checker.softAssert(mem_result == .ok, "memory leaks");

    checker.check();
}

File: test/_harnesses/rust/src/multiple.rs
// message files will include code that's not necessarily called in each test
#![allow(dead_code)]

use std::fs;

mod small_messages;

mod util;

use small_messages::MessageCodec;

fn main() {
    let mut checker = util::Checker { ok: true };
    let args = util::arg_parse();

    let empty_msg = small_messages::EmptyMessage {};
    let byte_msg = small_messages::ByteMessage { byte_member: 242 };
    let int_msg_a = small_messages::IntMessage { int_member: -42 };
    let int_msg_b = small_messages::IntMessage { int_member: 2048 };
    let float_msg = small_messages::FloatMessage { float_member: 1234.5678 };
    let long_msg = small_messages::LongMessage { int_member: 2147483647 + 10 };

    if args.contains_key("generate") {
        let mut writer: Vec<u8> = Vec::new();

        byte_msg.write_bytes(&mut writer, true);   //  0
        int_msg_a.write_bytes(&mut writer, true);  //  1
        int_msg_b.write_bytes(&mut writer, true);  //  2
        empty_msg.write_bytes(&mut writer, true);  //  3
        long_msg.write_bytes(&mut writer, true);   //  4
        float_msg.write_bytes(&mut writer, true);  //  5
        int_msg_a.write_bytes(&mut writer, true);  //  6
        int_msg_b.write_bytes(&mut writer, true);  //  7
        int_msg_b.write_bytes(&mut writer, true);  //  8
        int_msg_b.write_bytes(&mut writer, true);  //  9
        int_msg_a.write_bytes(&mut writer, true);  // 10
        empty_msg.write_bytes(&mut writer, true);  // 11

        let mut size = 0;
        size += byte_msg.get_size_in_bytes();
        size += int_msg_a.get_size_in_bytes() * 3;
        size += int_msg_b.get_size_in_bytes() * 4;
        size += empty_msg.get_size_in_bytes() * 2;
        size += long_msg.get_size_in_bytes();
        size += float_msg.get_size_in_bytes();
        size += 12;

        let filename = args.get("generate").unwrap();
        checker.soft_assert(writer.len() == size, "size calculation check");
        fs::write(filename, writer).unwrap();
    }
    else if args.contains_key("read") {
        let filename = args.get("read").unwrap();
        let mut buffer = fs::read(&filename).unwrap();
        buffer.resize(buffer.len() + 25, 0);
        let mut reader = small_messages::BufferReader::new(buffer);
        let msg_list = small_messages::process_raw_bytes(&mut reader, -1).unwrap();
        checker.soft_assert(msg_list.len() == 12, "reading multiple messages length");

        match &msg_list[0] {
            small_messages::Message::ByteMessage(bm) => checker.soft_assert(bm.byte_member == byte_msg.byte_member, "msg 0 content"),
            _ => checker.soft_assert(false, "msg 0 type"),
        }
        match &msg_list[1] {
            small_messages::Message::IntMessage(im) => checker.soft_assert(im.int_member == int_msg_a.int_member, "msg 1 content"),
            _ => checker.soft_assert(false, "msg 1 type"),
        }
        match &msg_list[2] {
            small_messages::Message::IntMessage(im) => checker.soft_assert(im.int_member == int_msg_b.int_member, "msg 2 content"),
            _ => checker.soft_assert(false, "msg 2 type"),
        }
        match &msg_list[3] {
            small_messages::Message::EmptyMessage(_em) => {},
            _ => checker.soft_assert(false, "msg 3 type"),
        }
        match &msg_list[4] {
            small_messages::Message::LongMessage(lm) => checker.soft_assert(lm.int_member == long_msg.int_member, "msg 4 content"),
            _ => checker.soft_assert(false, "msg 4 type"),
        }
        match &msg_list[5] {
            small_messages::Message::FloatMessage(fm) => checker.soft_assert(fm.float_member == float_msg.float_member, "msg 5 content"),
            _ => checker.soft_assert(false, "msg 5 type"),
        }
        match &msg_list[6] {
            small_messages::Message::IntMessage(im) => checker.soft_assert(im.int_member == int_msg_a.int_member, "msg 6 content"),
            _ => checker.soft_assert(false, "msg 6 type"),
        }
        match &msg_list[7] {
            small_messages::Message::IntMessage(im) => checker.soft_assert(im.int_member == int_msg_b.int_member, "msg 7 content"),
            _ => checker.soft_assert(false, "msg 7 type"),
        }
        match &msg_list[8] {
            small_messages::Message::IntMessage(im) => checker.soft_assert(im.int_member == int_msg_b.int_member, "msg 8 content"),
            _ => checker.soft_assert(false, "msg 8 type"),
        }
        match &msg_list[9] {
            small_messages::Message::IntMessage(im) => checker.soft_assert(im.int_member == int_msg_b.int_member, "msg 9 content"),
            _ => checker.soft_assert(false, "msg 9 type"),
        }
        match &msg_list[10] {
            small_messages::Message::IntMessage(im) => checker.soft_assert(im.int_member == int_msg_a.int_member, "msg 9 content"),
            _ => checker.soft_assert(false, "msg 9 type"),
        }
        match &msg_list[11] {
            small_messages::Message::EmptyMessage(_em) => {},
            _ => checker.soft_assert(false, "msg 11 type"),
        }
    }

    checker.check();
}
